### PR TITLE
Allow cilium to be used on seeds.

### LINKED ISSUE #4309: Allow usage of CNI with eBPF (Cilium, GKE Dataplane2) when using API SNI feature
**How to categorize this issue?**

/area cost
/area networking
/area control-plane
/kind enhancement
/priority 2

**What would you like to be added**:
GKE made the DataplaneV2 based on Cilium and eBPF generally available starting with K8s 1.20.6-gke1000 (https://cloud.google.com/kubernetes-engine/docs/concepts/dataplane-v2). Upgrade from existing clusters to the new dataplane are not supported.

If a GKE cluster with DataplaneV2 is used together with the Gardener ApiServer SNI feature, the dependency of the istio sidecar apparently prevents communication from the base clusters Gardenlet with a control plane apiserver for a new shoot. 

AFAIK this was already mentioned in the initial proposal for the SNI feature: https://github.com/gardener/gardener/blob/master/docs/proposals/08-shoot-apiserver-via-sni.md#future-work 

the same is true when running ManagedSeeds or Shoots with the APIServerSNI feature

**Why is this needed**:
To allow usage of GKE clusters with the new Dataplane V2

### PR DESCRIPTION
**Pull request by ScheererJ**
**How to categorize this PR?**

/area networking
/kind bug

**What this PR does / why we need it**:
Allow cilium to be used on seeds.

Kubernetes network policies are interpreted slightly different by different CNIs.
While calico works well with IP based policies, cilium expects IP ranges to be
cluster external. This is why the network policy, which should have made it possible
for the SNI istio-ingress gateway to forward traffic to a kube-apiserver in the
shoot control plane, was not working as expected causing traffic to be dropped.
Adding the istio-ingress gateway explicitly works well with cilium.

**Which issue(s) this PR fixes**:
Fixes #4309.

**Special notes for your reviewer**:

**Release note**:

```bugfix user
Allow cilium to be used on seeds with SNI enabled.
```

/cc @DockToFuture

### CODE REVIEW

File: pkg/operation/botanist/component/kubeapiserver/networkpolicies.go [hunk 1]
```diff
@@ -140,6 +141,14 @@ func (k *kubeAPIServer) reconcileNetworkPolicyAllowKubeAPIServer(ctx context.Con
 					From: []networkingv1.NetworkPolicyPeer{
 						// Allow all other Pods in the Seed cluster to access it.
 						{PodSelector: &metav1.LabelSelector{}},
+						// Allow all istio-ingress Pods in the Seed cluster to access it.
+						{
+							// we don't want to modify existing labels on the istio namespace
```
Comment (timebertt) [inline pkg/operation/botanist/component/kubeapiserver/networkpolicies.go:146]: > we don't want to modify existing labels on the istio namespace

Can you explain why?

File: pkg/operation/botanist/component/kubeapiserver/networkpolicies.go [hunk 2]
```diff
@@ -140,6 +141,14 @@ func (k *kubeAPIServer) reconcileNetworkPolicyAllowKubeAPIServer(ctx context.Con
...
+							NamespaceSelector: &metav1.LabelSelector{},
+							PodSelector: &metav1.LabelSelector{MatchLabels: map[string]string{
+								v1beta1constants.LabelApp: gardenletconfigv1alpha1.DefaultIngressGatewayAppLabelValue,
```
Comment (timebertt) [inline pkg/operation/botanist/component/kubeapiserver/networkpolicies.go:149]: Does this work as well, if additional exposure classes are configured in gardenlet?
Maybe we should add and use a common label, e.g. `istio=ingressgateway`.

**Review by timebertt — COMMENTED** [review]
Thank you for the PR!
I've added some questions inline :)

File: pkg/operation/botanist/component/kubeapiserver/networkpolicies.go (hunk 2 above)
Comment (ScheererJ) [reply pkg/operation/botanist/component/kubeapiserver/networkpolicies.go:149]: This is exactly the label that all exposure class deployments of istio have in common.

@dkistner: Feel free to confirm/deny.

File: pkg/operation/botanist/component/kubeapiserver/networkpolicies.go (hunk 1 above)
Comment (ScheererJ) [reply pkg/operation/botanist/component/kubeapiserver/networkpolicies.go:146]: Unfortunately, not really comprehensively.
The comment comes originally from https://github.com/gardener/gardener/pull/2406, i.e. the original SNI implementation. With this change it is now in four different places in the code base.
My assumption is that Martin did not want to change the istio labels to reduce the diff to the upstream release. However, it is now also not so straight forward anymore to change them as there are two places that create/manipulate the namespace. (Please note that this is mostly speculation on my side, though.)

File: pkg/operation/botanist/component/kubeapiserver/networkpolicies.go (hunk 2 above)
Comment (dkistner) [reply pkg/operation/botanist/component/kubeapiserver/networkpolicies.go:149]: If I remember correctly the only label that all ingress gateway pods (default and exposure class ones) have is `app: istio-ingressgateway`.

Comment (ScheererJ) [reply pkg/operation/botanist/component/kubeapiserver/networkpolicies.go:149]: (which is what I used here for that reason)

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**